### PR TITLE
rls: Implementation of lb config parse functionality.

### PR DESCRIPTION
**Pull request by easwars**

### CODE REVIEW

File: balancer/rls/internal/config.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+// MarshalJSON returns a JSON encoding of l.
+func (l *lbParsedConfig) MarshalJSON() ([]byte, error) {
+	return nil, fmt.Errorf("lbParsedConfig.MarshalJSON() is unimplemented")
+}
+
+// loadBalancingConfig represents a single load balancing config,
+// stored in JSON format.
+type loadBalancingConfig struct {
```
Comment (dfawley) [inline balancer/rls/internal/config.go]: We do similar things for load balancing config lists in https://github.com/grpc/grpc-go/blob/master/service_config.go#L252.  Maybe we should factor this out and re-use it?

File: balancer/rls/internal/builder.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	}
+
+	var lookupServiceTimeout, maxAge, staleAge time.Duration
+
+	if lst := cfg.rlsProto.GetLookupServiceTimeout(); lst != nil {
+		lookupServiceTimeout, err = ptypes.Duration(lst)
+		if err != nil {
+			return nil, fmt.Errorf("rls: failed to parse lookup_service_timeout in service config {%+v}", string(c))
+		}
+	}
```
Comment (dfawley) [inline balancer/rls/internal/builder.go]: Just an alternative to consider:
```go
  ...
  lookupServiceTimeout, err := convertDuration(cfg.rlsProto.GetLookupServiceTimeout())
  if err != nil {
    return nil, fmt.Errorf(...)
  }
  maxAge, err := convertDuration(cfg.rlsProto.GetMaxAge())
  // etc
  ...
}

func convertDuration(d *durpb.Duration) (time.Duration, error) {
  if d == nil {
    return 0, nil
  }
  return ptypes.Duration(d)
}
```

File: balancer/rls/internal/builder.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+	if resolver.Get(parsedTarget.Scheme) == nil {
+		return nil, fmt.Errorf("rls: invalid target URI in lookup_service {%s}", lookupService)
+	}
+
+	var lookupServiceTimeout, maxAge, staleAge time.Duration
+
+	if lst := cfg.rlsProto.GetLookupServiceTimeout(); lst != nil {
+		lookupServiceTimeout, err = ptypes.Duration(lst)
+		if err != nil {
+			return nil, fmt.Errorf("rls: failed to parse lookup_service_timeout in service config {%+v}", string(c))
```
Comment (dfawley) [inline balancer/rls/internal/builder.go]: Please include `err` in this error (and below).

File: balancer/rls/internal/builder.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+func (*rlsBB) Name() string {
+	return rlsBalancerName
+}
+
+// ParseConfig parses and validates the JSON representation of the service
+// config and returns the loadBalancingConfig to be used by the RLS LB policy.
+//
+// Helps implement the balancer.ConfigParser interface.
+//
+// The following validation checks are performed:
```
Comment (dfawley) [inline balancer/rls/internal/builder.go]: Would this validation logic be more appropriate in `config.go`?

File: balancer/rls/internal/builder.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+// * Marshals the map back into JSON, finds the config parser associated with
+//   the child policy and asks it to validate the config.
+// * If the validation succeeded, removes the dummy entry from the map and
+//   returns it. If any of the above steps failed, it returns an error.
+func validateChildPolicyConfig(cfg *lbParsedConfig) (map[string]json.RawMessage, error) {
+	var childConfig map[string]json.RawMessage
+	if err := json.Unmarshal(cfg.childPolicy.Config, &childConfig); err != nil {
+		return nil, fmt.Errorf("rls: json unmarshal failed for child policy config {%+v}: %v", cfg.childPolicy.Config, err)
+	}
+	childConfig[cfg.targetField], _ = json.Marshal(dummyChildPolicyTarget)
```
Comment (dfawley) [inline balancer/rls/internal/builder.go]: Do you mean to ignore the error here?

Comment (dfawley) [inline balancer/rls/internal/builder.go]: Is this supposed to be the `cpTargetField`?

File: balancer/rls/internal/config.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	cpName               string
+	cpTargetField        string
+	cpConfig             map[string]json.RawMessage
+}
+
+// lbParsedConfig is an parsed internal representation of the JSON
+// loadBalancing config for the RLS LB policy. This config is put through
+// further validation checks and transformed into another representation before
+// handing it off to the RLS LB policy.
+type lbParsedConfig struct {
```
[the remark this answers is not in the document]
Comment (easwars) [reply balancer/rls/internal/config.go]: Done. Ended up as a bigger refactor than I imagined. But looks better now I think.

**Comment by dfawley** [comment]
From travis/appengine:

`balancer/rls/internal/config_test.go:52:14: undefined: lbConfig`

**Comment by easwars** [comment]
Switched the build constraint to one based on Go version because that is really what we are constrained on here, because we need `strings.Builder` which was added only in `Go 1.10`

**Comment by easwars** [comment]
> Switched the build constraint to one based on Go version because that is really what we are constrained on here, because we need `strings.Builder` which was added only in `Go 1.10`

Travis seems happy now.

File: balancer/rls/internal/config.go [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+	childConfig[cpTargetField], _ = json.Marshal(dummyChildPolicyTarget)
+
+	jsonCfg, err := json.Marshal(childConfig)
+	if err != nil {
+		return nil, fmt.Errorf("rls: json marshal failed for child policy config {%+v}: %v", childConfig, err)
+	}
+	builder := balancer.Get(cp.Name)
+	if builder == nil {
+		// This should never happen since we already made sure that the child
+		// policy name mentioned in the service config is a valid one.
```
Comment (dfawley) [inline balancer/rls/internal/config.go]: Maybe we should store the child balancer.Builder itself (_instead_ of the name? or in addition to if necessary, although the `Name` method will give it to us if we need it for debugging) so we can be 100% sure it's impossible?

We could also validate it implements `ConfigParser` at that time, and store it in the config as `type interface BalancerConfigParser { balancer.Builder; balancer.ConfigParser }`.  (Maybe this is too crazy?)

...

I see now that this creates the `lbConfig` _after_ doing validation of the child policy, so the above might sound a little strange.  However, we'll soon have code that calls the child's `ParseConfig` to retrieve its updated config with the target name set, at which point that will make more sense.

Comment (easwars) [reply balancer/rls/internal/config.go]: Makes sense. Added a couple of TODOs for this.

**Pull request merged**